src/libethash/ethash.h

```diff
@@ -74,57 +84,143 @@ uint64_t ethash_get_cachesize(const uint32_t block_number);
+ * @param light    The light handler to free
+ */
+void ethash_light_delete(ethash_light_t light);
+/**
+ * Calculate the light client data
+ *
+ * @param ret            An object of ethash_return_value to hold the return value
+ * @param light          The light client handler
+ * @param params         The parameters to use
+ * @param header_hash    The header hash to pack into the mix
+ * @param nonce          The nonce to pack into the mix
+ * @return               true if all went well and false if there were invalid
+ *                       parameters given.
+ */
+bool ethash_light_compute(ethash_return_value *ret,
+                          ethash_light_t light,
+                          ethash_params const *params,
+                          const ethash_h256_t *header_hash,
+                          const uint64_t nonce);
+/**
+ * Get a pointer to the cache object held by the light client
+ *
+ * @param light    The light client whose cache to request
+ * @return         A pointer to the cache held by the light client or NULL if
+ *                 there was no cache in the first place
+ */
+ethash_cache *ethash_light_get_cache(ethash_light_t light);
+/**
+ * Move the memory ownership of the cache somewhere else
+ *
+ * @param light    The light client whose cache's memory ownership  to acquire.
+ *                 After this function concludes it will no longer have a cache.
+ * @return         A pointer to the moved cache or NULL if there was no cache in the first place
+ */
+ethash_cache *ethash_light_acquire_cache(ethash_light_t light);
 
-static inline void ethash_prep_full(void *full, ethash_params const *params, void const *cache)
-{
-    ethash_cache c;
-    c.mem = (void *) cache;
-    ethash_compute_full_data(full, params, &c);
-}
+/**
+ * Allocate and initialize a new ethash_full handler
+ *
+ * @param params    The parameters to initialize it with. We are interested in
+ *                  the full_size from here
+ * @param cache     A cache object to use that was allocated with @ref ethash_cache_new().
+ *                  Iff this function succeeds the ethash_full_t will take memory
+ *                  ownership of the cache and free it at deletion. If not then the user
+ *                  still has to handle freeing of the cache himself.
+ * @param seed      Block seedhash. TODO: Do we really need this in this function?
+ * @param callback  A callback function with signature of @ref ethash_callback_t
+ *                  It accepts an unsigned with which a progress of DAG calculation
+ *                  can be displayed. If all goes well the callback should return 0.
+ *                  If a non-zero value is returned then DAG generation will stop.
+ * @return          Newly allocated ethash_full handler or NULL in case of
+ *                  ERRNOMEM or invalid parameters used for @ref ethash_compute_full_data()
+ */
+ethash_full_t ethash_full_new(ethash_params const* params,
```

Comment:
I'm not sure it is the best idea to allocate memory for full cache inside the library. 
A user might want to write the cache to disk later. It would be easier if the user was the owner of the memory in first place. Consider mmap optimization.